source/native-plugins/midi-channel-ab.c

```diff
@@ -0,0 +1,217 @@
+} MidiGainHandle;
+
+// -----------------------------------------------------------------------
+
+static NativePluginHandle midichanab_instantiate(const NativeHostDescriptor* host)
+{
+    MidiGainHandle* const handle = (MidiGainHandle*)malloc(sizeof(MidiGainHandle));
+
+    if (handle == NULL)
+        return NULL;
+
+    handle->host = host;
+
+    for (int i=MAX_MIDI_CHANNELS; --i>=0;)
+        handle->channels[i] = 0.0f;
+
+    return handle;
+}
+
+#define handlePtr ((MidiGainHandle*)handle)
+
+static void midichanab_cleanup(NativePluginHandle handle)
+{
+    free(handlePtr);
+}
+
+static uint32_t midichanab_get_parameter_count(NativePluginHandle handle)
+{
+    return MAX_MIDI_CHANNELS;
+
+    // unused
+    (void)handle;
+}
+
+static const NativeParameter* midichanab_get_parameter_info(NativePluginHandle handle, uint32_t index)
+{
+    if (index > MAX_MIDI_CHANNELS)
+        return NULL;
+
+    static NativeParameter param;
+    static const NativeParameterScalePoint scalePoints[2] = { { "Output A", 0.0f }, { "Output B", 1.0f } };
+    static char paramName[24];
+
+    param.hints = NATIVE_PARAMETER_IS_ENABLED|NATIVE_PARAMETER_IS_AUTOMABLE|NATIVE_PARAMETER_IS_BOOLEAN|NATIVE_PARAMETER_USES_SCALEPOINTS;
+    param.name  = paramName;
+    param.unit  = NULL;
+    param.ranges.def       = 0.0f;
+    param.ranges.min       = 0.0f;
+    param.ranges.max       = 1.0f;
+    param.ranges.step      = 1.0f;
+    param.ranges.stepSmall = 1.0f;
+    param.ranges.stepLarge = 1.0f;
+    param.scalePointCount  = 2;
+    param.scalePoints      = scalePoints;
+
+    snprintf(paramName, 24, "%u", index+1);
+
+    return &param;
+
+    // unused
+    (void)handle;
+}
+
+static float midichanab_get_parameter_value(NativePluginHandle handle, uint32_t index)
+{
+    if (index > MAX_MIDI_CHANNELS)
+        return 0.0f;
+
+    return handlePtr->channels[index] ? 1.0f : 0.0f;
+}
+
+static void midichanab_set_parameter_value(NativePluginHandle handle, uint32_t index, float value)
+{
+    if (index > MAX_MIDI_CHANNELS)
+        return;
+
+    handlePtr->channels[index] = (value >= 0.5f);
+}
+
+static void midichanab_process(NativePluginHandle handle, float** inBuffer, float** outBuffer, uint32_t frames, const NativeMidiEvent* midiEvents, uint32_t midiEventCount)
+{
+    const NativeHostDescriptor* const host     = handlePtr->host;
+    const bool*                 const channels = handlePtr->channels;
+    NativeMidiEvent tmpEvent;
+
+    for (uint32_t i=0; i < midiEventCount; ++i)
+    {
+        const NativeMidiEvent* const midiEvent = &midiEvents[i];
+
+        const uint8_t status = (uint8_t)MIDI_GET_STATUS_FROM_DATA(midiEvent->data);
+
+        if (MIDI_IS_CHANNEL_MESSAGE(status))
+        {
+            const uint8_t channel = (uint8_t)MIDI_GET_CHANNEL_FROM_DATA(midiEvent->data);
+
+            if (channel >= MAX_MIDI_CHANNELS)
+                continue;
+
+            if (channels[channel])
+            {
+                tmpEvent.port    = midiEvent->port+1;
```

Comment:
this part is needlessly extensive.
please do a `memcpy(&tmpEvent, midiEvent, sizeof(NativeMidiEvent));` and then only modify the bits needed.